x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryBuilder.java

```diff
@@ -0,0 +1,125 @@
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * Parses and builds relevance_match queries
+ */
+public class RelevanceMatchQueryBuilder extends AbstractQueryBuilder<RelevanceMatchQueryBuilder> {
+
+    public static final String NAME = "relevance_match";
+    private static final Logger logger = LogManager.getLogger(RelevanceMatchQueryBuilder.class);
+    private static final ParseField FIELD_QUERY = new ParseField("query");
+
+    private static final ObjectParser<RelevanceMatchQueryBuilder, Void> PARSER = new ObjectParser<>(NAME, RelevanceMatchQueryBuilder::new);
+
+    static {
+        declareStandardFields(PARSER);
+
+        PARSER.declareString(RelevanceMatchQueryBuilder::setQuery, FIELD_QUERY);
+    }
+
+    private String query;
+
+    public RelevanceMatchQueryBuilder() {
+        super();
+    }
+
+    public RelevanceMatchQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+
+        query = in.readString();
+    }
+
+    public static RelevanceMatchQueryBuilder fromXContent(final XContentParser parser) {
+
+        final RelevanceMatchQueryBuilder builder;
+        try {
+            builder = PARSER.apply(parser, null);
+        } catch (final IllegalArgumentException e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+
+        if (builder.query == null) {
+            throw new ParsingException(parser.getTokenLocation(), "[relevance_match] requires a query, none specified");
+        }
+        return builder;
+    }
+
+    @Override
+    protected void doWriteTo(final StreamOutput out) throws IOException {
+        out.writeString(query);
+    }
+
+    @Override
+    protected void doXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startObject(NAME);
+
+        builder.field(FIELD_QUERY.getPreferredName(), query);
+
+        builder.endObject();
+    }
+
+    @Override
+    protected Query doToQuery(final SearchExecutionContext context) throws IOException {
+        // TODO add field retrieval (probably a holder class that gets populated with field mapping)
+        Collection<String> fields = QueryFieldsResolver.getQueryFields(context);
+        if (fields.isEmpty()) {
+            throw new IllegalArgumentException("[relevance_match] query cannot find text fields in the index");
+        }
+
+        final CombinedFieldsQueryBuilder builder = new CombinedFieldsQueryBuilder(query, fields.toArray(new String[fields.size()]));
+
+        final Query resultQuery = builder.toQuery(context);
+        logger.info("Result query: {}", resultQuery.toString());
```

Comment:
Would this make sense as a `DEBUG` level rather than an `INFO` level?

Reply:
Totally! At this stage I'd like to have visibility from the plugin logging, so it doesn't get buried with other Elasticsearch debug info.

I'll take a look into the logging config to activate debug for the plugin and change this. Thanks! 👍